### PR TITLE
Sample logs based on the configured SamplingRatio for traces

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Enable sampling for logs(Trace and Exception telemetry) based on the configured `SampleRatio` for traces. Sampling will only be enabled for logs that are emitted within an activity context.

### CODE REVIEW

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/OpenTelemetryBuilderExtensions.cs [hunk 1]
```diff
@@ -158,28 +156,27 @@ public static OpenTelemetryBuilder UseAzureMonitor(this OpenTelemetryBuilder bui
...
+
+                                // Sampling is disabled by the user.
+                                // Set the SampleRatio of log exporter to 1.0.
+                                // The value will be used by the exporter to set SampleRate on the telemetry items.
+                                o.SamplingRatio = 1.0f;
+                            });
                         }
 
+                        // Sequence of adding processors does not matter here even if the filtering processor is used.
+                        // LogRecord will pass through each processor independent of other processors in the pipeline.
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/OpenTelemetryBuilderExtensions.cs:179]: nit: I think this comment is out of place. It describes LogRecord filtering processor, but it's listed before the EnableLiveMetrics. Should this instead be moved to line 163?

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/CHANGELOG.md [hunk 2]
```diff
@@ -9,6 +9,8 @@
   ([#42174](https://github.com/Azure/azure-sdk-for-net/pull/42174))
 * Turned off internal spans and logs in exporter HTTP pipeline
   ([#43359](https://github.com/Azure/azure-sdk-for-net/pull/43359))
+* Logs emitted within an activity context will now be sampled by default based on the `SamplingRatio` configured. To disable this sampling set `DisableTraceBasedSamplingForLogs` to `true`.
+  ([#43793](https://github.com/Azure/azure-sdk-for-net/pull/43793))
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/CHANGELOG.md:13]:
```suggestion
* Logs emitted within an Activity context will now be sampled by default based on the `SamplingRatio` configured for Traces. To disable this sampling set `DisableTraceBasedSamplingForLogs` to `true`.
  ([#43793](https://github.com/Azure/azure-sdk-for-net/pull/43793))
```

**Comment by TimothyMothra** [comment]
Are you planning to update the Readme in this PR or in a follow up?

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Azure.Monitor.OpenTelemetry.AspNetCore.csproj [hunk 3]
```diff
@@ -24,10 +24,10 @@
     <!-- Depending on monthly deliverables, we may switch between PackageReference or ProjectReference. Keeping both here to make the switch easier. -->
 
     <!-- FOR PUBLIC RELEASES, MUST USE PackageReference. THIS REQUIRES A STAGGERED RELEASE IF SHIPPING A NEW EXPORTER. -->
-    <PackageReference Include="Azure.Monitor.OpenTelemetry.Exporter" />
+    <!--<PackageReference Include="Azure.Monitor.OpenTelemetry.Exporter" />-->
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Azure.Monitor.OpenTelemetry.AspNetCore.csproj:27]: Could you please revert the changes to this file.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/AzureMonitorOptions.cs [hunk 4]
```diff
@@ -58,6 +58,15 @@ public class AzureMonitorOptions : ClientOptions
         /// </summary>
         public string StorageDirectory { get; set; }
 
+        /// <summary>
+        /// Disables sampling for logs that are emitted within an activity context.
+        /// </summary>
+        /// <remarks>
+        /// By default, the logs that are emitted within an activity context are sampled
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/AzureMonitorOptions.cs:65]: Please call out explicitly, default is false.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Azure.Monitor.OpenTelemetry.AspNetCore.csproj (hunk 3 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/Azure.Monitor.OpenTelemetry.AspNetCore.csproj:27]: Without this, the tests won't work. We can revert it when we have an updated exporter pakage with the changes.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/OpenTelemetryBuilderExtensions.cs [hunk 5]
```diff
@@ -158,28 +156,27 @@ public static OpenTelemetryBuilder UseAzureMonitor(this OpenTelemetryBuilder bui
...
                         else
                         {
-                            loggingOptions.AddAzureMonitorLogExporter(o => azureMonitorOptions.SetValueToExporterOptions(o));
+                            loggingOptions.AddAzureMonitorLogExporter(o =>
+                            {
+                                azureMonitorOptions.SetValueToExporterOptions(o);
+
+                                // Sampling is disabled by the user.
+                                // Set the SampleRatio of log exporter to 1.0.
+                                // The value will be used by the exporter to set SampleRate on the telemetry items.
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/src/OpenTelemetryBuilderExtensions.cs:175]:
```suggestion
                                // The value will be used by the log exporter to set SampleRate on the telemetry items.
```

**Comment by vishweshbankwar** [comment]
> Are you planning to update the Readme in this PR or in a follow up?

Pushed the readme update as well.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Tests/E2EAzureMonitorDistroTests.cs [hunk 6]
```diff
@@ -80,42 +80,48 @@ public async Task ValidateTelemetryExport()
...
 
-                var sp = sv.BuildServiceProvider();
-                var loggerProvider = sp.GetRequiredService<ILoggerProvider>();
-                var sdkProvider = typeof(OpenTelemetryLoggerProvider).GetField("Provider", BindingFlags.Instance | BindingFlags.NonPublic)?.GetValue(loggerProvider);
-                var processor = sdkProvider?.GetType().GetProperty("Processor", BindingFlags.Instance | BindingFlags.Public)?.GetMethod?.Invoke(sdkProvider, null);
+            var compositeProcessor = processor as CompositeProcessor<LogRecord>;
 
-                Assert.NotNull(processor);
+            // Filtering processor/ exporter is added first in the pipeline.
+            var firstProcessor = typeof(CompositeProcessor<LogRecord>).GetField("Head", BindingFlags.Instance | BindingFlags.NonPublic)?.GetValue(compositeProcessor);
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Tests/E2EAzureMonitorDistroTests.cs:106]: Not a blocker for this PR, this test pattern has an issue if we add another process or rearrange it becomes very difficult to modify this test and need to create an alternate one. Below is the one of the test that was skipped earlier to merge live metrics log changes. 

https://github.com/rajkumar-rangaraj/azure-sdk-for-net/blob/9246a13965898cb304c305e3bbd923d1de31057e/sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Tests/E2EAzureMonitorDistroTests.cs#L83

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/CHANGELOG.md [hunk 7]
```diff
@@ -9,6 +9,8 @@
   ([#42174](https://github.com/Azure/azure-sdk-for-net/pull/42174))
 * Turned off internal spans and logs in exporter HTTP pipeline
   ([#43359](https://github.com/Azure/azure-sdk-for-net/pull/43359))
+* Logs emitted within an Activity context will now be sampled by default based on the `SamplingRatio` configured for Traces. To disable this sampling set `DisableTraceBasedSamplingForLogs` to `true`.
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/CHANGELOG.md:12]: Is there a public api for exporter?

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/AzureMonitorLogExporter.cs [hunk 8]
```diff
@@ -19,18 +19,20 @@ public sealed class AzureMonitorLogExporter : BaseExporter<LogRecord>
...
         /// <param name="options">Configuration options for the exporter.</param>
-        public AzureMonitorLogExporter(AzureMonitorExporterOptions options) : this(TransmitterFactory.Instance.Get(options))
+        public AzureMonitorLogExporter(AzureMonitorExporterOptions options) : this(options, TransmitterFactory.Instance.Get(options))
         {
         }
 
-        internal AzureMonitorLogExporter(ITransmitter transmitter)
+        internal AzureMonitorLogExporter(AzureMonitorExporterOptions options, ITransmitter transmitter)
         {
+            _sampleRate = (float)Math.Round(options.SamplingRatio * 100);
```
Comment (rajkumar-rangaraj) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/AzureMonitorLogExporter.cs:35]: can we check if distro and set samplerate? In the other part of the code we could just check null on samplerate.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Tests/E2EAzureMonitorDistroTests.cs (hunk 6 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Tests/E2EAzureMonitorDistroTests.cs:106]: Yeah, we need better e2e tests. But I would still keep this. In general, we should not be changing the sequence of processors. This can catch issues early if in future we depend on certain processor sequence.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/CHANGELOG.md (hunk 7 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/CHANGELOG.md]: My bad. This should be in distro. Fixed.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/AzureMonitorLogExporter.cs (hunk 8 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/AzureMonitorLogExporter.cs:35]: `isDistro` may not be initialized during ctor execution. It is set after the first export when resource is accessed.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md [hunk 9]
```diff
@@ -133,6 +133,20 @@ builder.Services.AddOpenTelemetry().UseAzureMonitor(o =>
...
 ```
 
+#### Disable Logs Sampling
+
+By default, logs emitted within an Activity context are sampled based on the configured [sampling ratio](#customizing-sampling-percentage) for traces. This sampling can be disabled as shown below.
+
+``` C#
+builder.Services.AddOpenTelemetry().UseAzureMonitor(o =>
+{
+    o.DisableTraceBasedSamplingForLogs = true;
```
Comment (reyang) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md:143]: I have a general question regarding how to organize the Boolean options, do we want a mixure of "EnableXyz" and "DisableXyz", or it'll be better to stick with "EnableXyz"?

For example, "EnableTraceBasedSamplingForLogs = false".

Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/README.md:143]: For this particular name, it is driven by the specification consistent with all languages.

But in general, the choice is based on

1) Name should reflect the intent clearly.
2) Consistency with other options when possible.

Disable makes the intent more clear than enable with default `true`. Right now, we are deviating on `EnableLiveMetrics = true` which I think should be changed to Disable as well.

**Comment by TimothyMothra** [comment]
@vishweshbankwar builds are failing due to link validation. 
we're referencing instrumentation libraries that were migrated to the contrib repo.
I opened a new PR to update our Readme: https://github.com/Azure/azure-sdk-for-net/pull/43803
